File: lib_openweather/src/main/java/guepardoapps/library/openweather/controller/BroadcastController.java
package guepardoapps.library.openweather.controller;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.annotation.NonNull;

import java.io.Serializable;
import java.util.Locale;

import guepardoapps.library.openweather.utils.Logger;

@SuppressWarnings({"WeakerAccess"})
public class BroadcastController implements IBroadcastController {
    private static String Tag = BroadcastController.class.getSimpleName();

    private Context _context;

    public BroadcastController(@NonNull Context context) {
        _context = context;
    }

    @Override
    public void SendSimpleBroadcast(@NonNull String broadcast) {
        Intent broadcastIntent = new Intent(broadcast);
        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendIntBroadcast(@NonNull String broadcast, @NonNull String bundleName, int data) {
        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        broadcastData.putInt(bundleName, data);
        broadcastIntent.putExtras(broadcastData);
        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendIntParcelableBroadcast(@NonNull String broadcast, @NonNull String[] bundleNamesInteger, @NonNull int[] dataInteger, @NonNull String[] bundleNamesParcelable, @NonNull Object[] dataParcelables) {
        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        for (int index = 0; index < bundleNamesInteger.length; index++) {
            broadcastData.putInt(bundleNamesInteger[index], dataInteger[index]);
        }
        for (int index = 0; index < bundleNamesParcelable.length; index++) {
            broadcastData.putParcelable(bundleNamesParcelable[index], (Parcelable) dataParcelables[index]);
        }
        broadcastIntent.putExtras(broadcastData);
        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendIntArrayBroadcast(@NonNull String broadcast, @NonNull String[] bundleNames, @NonNull int[] data) {
        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        for (int index = 0; index < bundleNames.length; index++) {
            broadcastData.putInt(bundleNames[index], data[index]);
        }
        broadcastIntent.putExtras(broadcastData);
        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendDoubleBroadcast(@NonNull String broadcast, @NonNull String bundleName, double data) {
        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        broadcastData.putDouble(bundleName, data);
        broadcastIntent.putExtras(broadcastData);
        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendBooleanBroadcast(@NonNull String broadcast, @NonNull String bundleName, boolean data) {
        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        broadcastData.putBoolean(bundleName, data);
        broadcastIntent.putExtras(broadcastData);
        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendStringBroadcast(@NonNull String broadcast, @NonNull String bundleName, @NonNull String data) {
        if (data.length() >= 1024) {
            Logger.getInstance().Error(Tag, String.format(Locale.getDefault(), "Invalid data length %d!", data.length()));
            return;
        }

        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        broadcastData.putString(bundleName, data);
        broadcastIntent.putExtras(broadcastData);

        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendStringArrayBroadcast(@NonNull String broadcast, @NonNull String[] bundleNames, @NonNull String[] data) {
        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        for (int index = 0; index < bundleNames.length; index++) {
            broadcastData.putString(bundleNames[index], data[index]);
        }
        broadcastIntent.putExtras(broadcastData);
        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendSerializableBroadcast(@NonNull String broadcast, @NonNull String bundleName, @NonNull Object model) {
        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        broadcastData.putSerializable(bundleName, (Serializable) model);
        broadcastIntent.putExtras(broadcastData);
        _context.sendBroadcast(broadcastIntent);
    }

    @Override
    public void SendSerializableArrayBroadcast(@NonNull String broadcast, @NonNull String[] bundleNames, @NonNull Object[] models) {
        if (bundleNames.length != models.length) {
            Logger.getInstance().Warning(Tag, "Cannot send broadcast! length are not equal!");
            return;
        }

        Intent broadcastIntent = new Intent(broadcast);
        Bundle broadcastData = new Bundle();
        for (int index = 0; index < bundleNames.length; index++) {
            broadcastData.putSerializable(bundleNames[index], (Serializable) models[index]);
        }
        broadcastIntent.putExtras(broadcastData);
        _context.sendBroadcast(broadcastIntent);
    }
}

File: app/src/main/java/guepardoapps/lib/openweather/controller/BroadcastController.kt
package guepardoapps.lib.openweather.controller

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.annotation.NonNull
import guepardoapps.lib.openweather.utils.Logger
import java.io.Serializable

class BroadcastController(@NonNull private val context: Context) : IBroadcastController {

    private val tag: String = BroadcastController::class.java.canonicalName

    override fun simpleBroadcast(broadcast: String) {
        context.sendBroadcast(Intent(broadcast))
    }

    override fun boolBroadcast(broadcast: String, key: String, data: Boolean) {
        val bundle = Bundle()
        bundle.putBoolean(key, data)
        val intent = Intent(broadcast)
        intent.putExtras(bundle)
        context.sendBroadcast(intent)
    }

    override fun intBroadcast(broadcast: String, key: String, data: Int) {
        val bundle = Bundle()
        bundle.putInt(key, data)
        val intent = Intent(broadcast)
        intent.putExtras(bundle)
        context.sendBroadcast(intent)
    }

    override fun doubleBroadcast(broadcast: String, key: String, data: Double) {
        val bundle = Bundle()
        bundle.putDouble(key, data)
        val intent = Intent(broadcast)
        intent.putExtras(bundle)
        context.sendBroadcast(intent)
    }

    override fun stringBroadcast(broadcast: String, key: String, data: String) {
        val bundle = Bundle()
        bundle.putString(key, data)
        val intent = Intent(broadcast)
        intent.putExtras(bundle)
        context.sendBroadcast(intent)
    }

    override fun serializableBroadcast(broadcast: String, key: String, data: Serializable) {
        val bundle = Bundle()
        bundle.putSerializable(key, data)
        val intent = Intent(broadcast)
        intent.putExtras(bundle)
        context.sendBroadcast(intent)
    }

    override fun boolArrayBroadcast(broadcast: String, keyArray: Array<String>, dataArray: Array<Boolean>) {
        if (keyArray.size != dataArray.size) {
            Logger.instance.error(tag, "boolArrayBroadcast: Incompatible array size: keyArray: ${keyArray.size}, dataArray: ${dataArray.size}")
            return
        }

        val bundle = Bundle()
        for (index in 0 until keyArray.size step 1) {
            bundle.putBoolean(keyArray[index], dataArray[index])
        }

        val intent = Intent(broadcast)
        intent.putExtras(bundle)

        context.sendBroadcast(intent)
    }

    override fun intArrayBroadcast(broadcast: String, keyArray: Array<String>, dataArray: Array<Int>) {
        if (keyArray.size != dataArray.size) {
            Logger.instance.error(tag, "intArrayBroadcast: Incompatible array size: keyArray: ${keyArray.size}, dataArray: ${dataArray.size}")
            return
        }

        val bundle = Bundle()
        for (index in 0 until keyArray.size step 1) {
            bundle.putInt(keyArray[index], dataArray[index])
        }

        val intent = Intent(broadcast)
        intent.putExtras(bundle)

        context.sendBroadcast(intent)
    }

    override fun doubleArrayBroadcast(broadcast: String, keyArray: Array<String>, dataArray: Array<Double>) {
        if (keyArray.size != dataArray.size) {
            Logger.instance.error(tag, "doubleArrayBroadcast: Incompatible array size: keyArray: ${keyArray.size}, dataArray: ${dataArray.size}")
            return
        }

        val bundle = Bundle()
        for (index in 0 until keyArray.size step 1) {
            bundle.putDouble(keyArray[index], dataArray[index])
        }

        val intent = Intent(broadcast)
        intent.putExtras(bundle)

        context.sendBroadcast(intent)
    }

    override fun stringArrayBroadcast(broadcast: String, keyArray: Array<String>, dataArray: Array<String>) {
        if (keyArray.size != dataArray.size) {
            Logger.instance.error(tag, "stringArrayBroadcast: Incompatible array size: keyArray: ${keyArray.size}, dataArray: ${dataArray.size}")
            return
        }

        val bundle = Bundle()
        for (index in 0 until keyArray.size step 1) {
            bundle.putString(keyArray[index], dataArray[index])
        }

        val intent = Intent(broadcast)
        intent.putExtras(bundle)

        context.sendBroadcast(intent)
    }

    override fun serializableArrayBroadcast(broadcast: String, keyArray: Array<String>, dataArray: Array<Serializable>) {
        if (keyArray.size != dataArray.size) {
            Logger.instance.error(tag, "serializableArrayBroadcast: Incompatible array size: keyArray: ${keyArray.size}, dataArray: ${dataArray.size}")
            return
        }

        val bundle = Bundle()
        for (index in 0 until keyArray.size step 1) {
            bundle.putSerializable(keyArray[index], dataArray[index])
        }

        val intent = Intent(broadcast)
        intent.putExtras(bundle)

        context.sendBroadcast(intent)
    }
}